C/services/south/south.cpp

```diff
@@ -673,3 +711,61 @@ int SouthService::createTimerFd(struct timeval rate)
 	return fd;
 }
 
+/**
+ * If enabled, control the throttling of the poll rate in order to keep
+ * the buffer usage of the service within check.
+ *
+ * Although this is written as if rate is being control, which it
+ * logically is, the actual values are poll intervals. Hence reducing
+ * the poll rate increases the value of m_currentRate.
+ */
+void SouthService::throttlePoll()
+{
+struct timeval now, res;
+
+	if (!m_throttle)
+	{
+		return;
+	}
+	double desired = m_desiredRate.tv_sec + ((double)m_desiredRate.tv_usec / 1000000);
+	gettimeofday(&now, NULL);
+	timersub(&now, &m_lastThrottle, &res);
+	if (m_ingest->queueLength() > m_highWater && res.tv_sec > SOUTH_THROTTLE_DOWN_INTERVAL)
+	{
+		double rate = m_currentRate.tv_sec + ((double)m_currentRate.tv_usec / 1000000);
+		rate *= (1.0 + ((double)SOUTH_THROTTLE_PERCENT / 100.0));
+		m_currentRate.tv_sec = (long)rate;
+		m_currentRate.tv_usec = (rate - m_currentRate.tv_sec) * 1000000;
```

Comment:
If rate is initially say 1.9 secs, then rate becomes 2.09. Then at line 738, m_currentRate.tv_usec becomes more than 1 sec. There may be a need to normalize m_currentRate structure.